### PR TITLE
feat: record and replay any remote backend

### PR DESCRIPTION
**Pull request by riccardo-forina**
**⚠️ TECH PREVIEW ⚠️**

This PR introduces the capability for the `syndesis` dev server to interact with any backend - local or remote - through a [proxy set up by CRA](https://facebook.github.io/create-react-app/docs/proxying-api-requests-in-development) and record the responses to be able to replay them in a later moment (eg. for testing). 

This effectively removes the need to have the whole syndesis dev environment installed locally for the most common development needs.

There are new scripts available to run:
* `BACKEND=http://... yarn proxy` runs the app in the development mode. API calls will be proxied to the provided `BACKEND` URL.
* `record` works like `proxy`, but requests and responses will be recorded in the `tapes` folder. This command should be used only to record Cypress test sessions.
* `replay` runs the app in the development mode. API calls will be matched against the available recordings. No actual backend call will be performed. This command should be used only to run Cypress tests.

## How `proxy` works

* a hand-crafted `config.json`  that sets `apiBase` to be the same as the development server address is set up; this allows us to intercept requests to `/api`
* ask the user to login against the provided `BACKEND` to retrieve the right Cookies to use with the proxy. This is done by a Chromium instance controlled by Puppeteer through the `syndesis/scripts/get-login-cookie.js` script.
* pass the cookie to the proxy and fire an instance of [talkback](https://github.com/ijpiantanida/talkback) (that's the underlying tool that does the actual proxying, recording and replaying of the requests) running on `localhost:8556`
* record the requests done while using the app

Recordings will be stored in `syndesis/tapes/[value of the syndesis-mock-session request header]full/api/path/xxx.json5`.

## How to replay a session

```bash
yarn replay
``` 

This command will:
* a hand-crafted `config.json`  that sets `apiBase` to be the same as the development server address is set up; this allows us to intercept requests to `/api`
* will use `syndesis/tapes` as the tapes source
* requests will be matched against the tapes by matching the URL, verb and some headers (including the syndesis-mock-session one, which allows recording multiple versions of the same response tying it to the test suite that performed the request).

### CODE REVIEW

File: syndesis/src/setupProxy.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+const config = require('../public/config.json');
+
+// Disable self-signed certificate check
+process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';
+
+const backendName = filenamify(process.env.BACKEND);
+
+const server = talkback({
+  host: process.env.BACKEND,
+  port: 556,
```
Comment (gashcrumb) [inline syndesis/src/setupProxy.js]: Can we pick a port > 1024 for this?  On Linux at least you can't open ports < 1024 unless you're root.

**Comment by gashcrumb** [comment]
Taking the time to finally review this and try it out since I'm thinking of borrowing some stuff :-)

File: syndesis/src/setupProxy.js (hunk 1 above)
Comment (gashcrumb) [reply syndesis/src/setupProxy.js]: Tryin' 5556 locally for now.

File: syndesis/src/setupProxy.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  ],
+  summary: false,
+  debug: true,
+});
+server.start(() => console.log('Talkback Started'));
+
+module.exports = function(app) {
+  app.use(
+    proxy('/api', {
+      target: 'http://localhost:556',
```
Comment (gashcrumb) [inline syndesis/src/setupProxy.js]: Then this line would have to change also...

File: syndesis/src/setupProxy.js (hunk 1 above)
Comment (riccardo-forina) [reply syndesis/src/setupProxy.js]: Ah good point, yes it can be whatever we want it to be!

**Comment by gashcrumb** [comment]
Also when running `yarn start:mocks:record` (I dunno why I foolishly decided to first run `yarn start:mocks:replay :smile: ) am I supposed to need to close the browser window after logging in and getting redirected to staging?  If I let it sit there I see a timeout exception in the script output, same behavior if I close it, and then in either case I see that all the backend requests are getting oauth redirect responses rather than the proper API response.

**Comment by riccardo-forina** [comment]
> Also when running `yarn start:mocks:record` (I dunno why I foolishly decided to first run `yarn start:mocks:replay 😄 ) am I supposed to need to close the browser window after logging in and getting redirected to staging? If I let it sit there I see a timeout exception in the script output, same behavior if I close it, and then in either case I see that all the backend requests are getting oauth redirect responses rather than the proper API response.

That window should close on its own. If it doesn't then the script that intercepts the cookie isn't working properly. Just wondering, is the `BACKEND` url you provided with or without an ending `/`? Because I found it should not have the ending `/` to work.

**Comment by gashcrumb** [comment]
Yep, that sorted it, thanks @riccardo-forina!

**Comment by riccardo-forina** [comment]
phew 😂 yeah that behavior should be documented, or the script should warn about this if a trailing `/` is found

**Comment by gashcrumb** [comment]
I ain't complaining too much, it's awesome seeing it working :-)  So +1 for merging once the port's changed.

**Comment by riccardo-forina** [comment]
New commands:

`BACKEND=https://syndesis.192.168.64.1.nip.io yarn start:proxy` to start a proxy to `BACKEND` without any recording or replay thingy.

`BACKEND=https://syndesis.192.168.64.1.nip.io yarn record session-name` to start a proxy to `BACKEND` and recording all requests/responses in a folder named `session-name`

`yarn replay session-name` to proxy all requests to the mock server in replay-only mode, using the previously recorded session `session-name` to load the tapes.

WIP: figure out how to set up the app in jarn tests so to fire up the mock server on demand pointing to a previously recorded session, to simulate e2e testing.

**Comment by riccardo-forina** [comment]
The mock server is also available as a stand alone script in ` ./scripts/mock-server.js`

```
./scripts/mock-server.js           
mock-server.js replay <session>

Replay the previously recorded <session>

Positionals:
  session

Options:
  --version         Show version number                                [boolean]
  --port            This server port.                   [number] [default: 8556]
  --recordingsPath  Path where to load and save recorded sessions.
                                                              [string] [default:
                 "/Users/riccardoforina/Code/syndesis-react/dev/syndesis/tapes"]
  --silent          Disable requests information console messages in the middle
                    of requests.                                [default: false]
  --summary         Enable exit summary of new and unused tapes at exit.
                                                                [default: false]
  --debug           Enable verbose debug information.           [default: false]
  --help            Show help                                          [boolean]

Examples:
  mock-server.js dashboard  Replays a previously recorded session named
                            dashboard

Not enough non-option arguments: got 0, need at least 1
```

```
./scripts/mock-server.js record
mock-server.js record <session> <host>

Record a <session> proxying request to <host>. If <session> already exists, it
will be updated.

Positionals:
  session                                                             [required]
  host     The remote backend host                                    [required]

Options:
  --version         Show version number                                [boolean]
  --port            This server port.                   [number] [default: 8556]
  --recordingsPath  Path where to load and save recorded sessions.
                                                              [string] [default:
                 "/Users/riccardoforina/Code/syndesis-react/dev/syndesis/tapes"]
  --silent          Disable requests information console messages in the middle
                    of requests.                                [default: false]
  --summary         Enable exit summary of new and unused tapes at exit.
                                                                [default: false]
  --debug           Enable verbose debug information.           [default: false]
  --help            Show help                                          [boolean]

Examples:
  mock-server.js record dashboard           Replays a previously recorded
  https://syndesis.192.168.64.1.nip.io      session named dashboard

Not enough non-option arguments: got 0, need at least 2
```

**Comment by riccardo-forina** [comment]
PR Storybook available [here](https://circleci.com/api/v1.1/project/github/syndesisio/syndesis-react-poc/356/artifacts/0/home/circleci/project/doc/index.html)

**Comment by riccardo-forina** [comment]
PR Storybook available [here](https://circleci.com/api/v1.1/project/github/syndesisio/syndesis-react-poc/360/artifacts/0/home/circleci/project/doc/index.html)

**Comment by riccardo-forina** [comment]
PR Storybook available [here](https://circleci.com/api/v1.1/project/github/syndesisio/syndesis-react-poc/363/artifacts/0/home/circleci/project/doc/index.html)

**Comment by riccardo-forina** [comment]
PR Storybook available [here](https://circleci.com/api/v1.1/project/github/syndesisio/syndesis-react-poc/365/artifacts/0/home/circleci/project/doc/index.html)

**Comment by riccardo-forina** [comment]
PR Storybook available [here](https://circleci.com/api/v1.1/project/github/syndesisio/syndesis-react-poc/407/artifacts/0/home/circleci/project/doc/index.html)

**Comment by riccardo-forina** [comment]
This should hopefully be in better shape now.

* storybook for the app is gone.
* Jest for the app should be usable.
* the mocks recorder/replayer has its own CLI, to hopefully make it easier to work with it.
* Cypress for integration testing is now available for the app. There is a single test right now to showcase how it works; to see it working you'll need `yarn replay` running, and then in another shell `yarn run cypress open`

Cypress is in "preview mode", it's not integrated to CI and I don't expect us doing that anytime soon. At least not until we have a critical mass of Cypress tests, which means that we have eaten our own dogfood and fixed the millions of bugs I introduced with this.

**Comment by riccardo-forina** [comment]
Oh, I also reworked the README a bit. Any feedback about it is appreciated!

**Pull request merged**